server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -348,61 +349,68 @@ protected void doInternalExecute(Task task, BulkRequest bulkRequest, Executor ex
-            // We should only auto create if we are not requiring it to be an alias
-            .filter(entry -> entry.getValue().isRequireAlias == false)
-            .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().isRequireDataStream));
-
-        // Step 3: Collect all the data streams that need to be rolled over before writing
-        Set<String> dataStreamsToBeRolledOver = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER)
-            ? indices.keySet().stream().filter(target -> {
-                DataStream dataStream = state.metadata().dataStreams().get(target);
-                return dataStream != null && dataStream.rolloverOnWrite();
-            }).collect(Collectors.toSet())
-            : Set.of();
-
-        // Step 4: create all the indices that are missing, if there are any missing. start the bulk after all the creates come back.
         createMissingIndicesAndIndexData(
             task,
             bulkRequest,
             executor,
             listener,
             indicesToAutoCreate,
             dataStreamsToBeRolledOver,
-            indicesThatCannotBeCreated,
+            failureStoresToBeRolledOver,
             startTime
         );
     }
 
-    /*
-     * This method is responsible for creating any missing indices, rolling over a data stream when needed and then
-     *  indexing the data in the BulkRequest
+    private void determineMissingTargets(
+        BulkRequest bulkRequest,
+        Map<String, Boolean> indicesToAutoCreate,
+        Set<String> dataStreamsToBeRolledOver,
+        Set<String> failureStoresToBeRolledOver
+    ) {
+        Map<String, Boolean> indexExistence = new HashMap<>();
+        ClusterState state = clusterService.state();
+        boolean lazyRolloverFeature = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER);
+
+        for (DocWriteRequest<?> request : bulkRequest.requests) {
+            if (request.opType() == OpType.DELETE
+                && request.versionType() != VersionType.EXTERNAL
+                && request.versionType() != VersionType.EXTERNAL_GTE) {
+                continue;
+            }
+            Boolean indexExists = indexExistence.get(request.index());
+            if (indexExists == null) {
+                indexExists = indexNameExpressionResolver.hasIndexAbstraction(request.index(), state);
+                indexExistence.put(request.index(), indexExists);
+            }
+            if (indexExists == false && request.isRequireAlias() == false) {
+                Boolean requiresDataStream = indicesToAutoCreate.get(request.index());
+                if (requiresDataStream == null || (requiresDataStream == false && request.isRequireDataStream())) {
+                    indicesToAutoCreate.put(request.index(), request.isRequireDataStream());
+                }
```

Comment:
I don't think I can extract a lambda to outside the loop in this case. Can you explain what you don't like about "tri-state booleans"? I think it's a pretty common pattern to use an `== null` check to determine if a key is present in a map (as it avoids an extra `containsKey` call). Do you have any suggestions that don't require creating a new lambda inside the loop?

Reply:
Usually a `boolean` has two values, true or false. When using `Boolean` you also have the option to have `null`. So there are 3 possible values, sometimes, null just defaults to true or false, but in this case it means something else, it means what `null` usually means, I do not have a value for this and there is a different action you need to take. 

I can see how using the `contains` does not help with readability either because without `requiresDataStream = indicesToAutoCreate.get(request.index())` it's hard to remember what does that boolean represent.